### PR TITLE
Drop support for Python 3.7

### PR DESCRIPTION
**Pull request by cdce8p**
## Description
Similar to https://github.com/pylint-dev/pylint/pull/8609
EOL for Python 3.7 is 2023-06-27.

`2.16.0` already includes a lot of (unintentional) breaking changes. We might as well include this one too.
_Maybe even do `3.0.0` instead?_

### CODE REVIEW

**Comment by DanielNoord** [comment]
I'd be in favour of `3.0` and dropping `3.7`. However, that would also mean that `pylint` `3.x` would be `3.8+`. I'll leave that decision to @Pierre-Sassoulas  but I am in favour!

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
I wanted to support 3.7 until the end, but on the other hand releasing astroid/pylint 3.0 and then immediately removing support is not ideal either (the last version could be half broken). flake8 dropped 3.7 months ago, but ruff is still supporting it. Would dropping python 3.7 right now would make our life that much easier ?

**Comment by DanielNoord** [comment]
> Would dropping python 3.7 right now would make our life that much easier ?

It removes a dependency and a lot of if statements.

We can also require `end_lineno` to be `int` for a lot more nodes, which feels like a good thing considering all the breaking changes to constructors. I'd say it is a definite improvement.

**Comment by Pierre-Sassoulas** [comment]
Releasing astroid 3.0 make sense too. The lambda/function def change is impactful and does break compatibility with old version of pylint.

File: astroid/__init__.py [hunk 1]
```diff
@@ -181,7 +181,7 @@
 from astroid.util import Uninferable
 
 # Performance hack for tokenize. See https://bugs.python.org/issue43014
-# Adapted from https://github.com/pylint-dev/pycodestyle/pull/993
+# Adapted from https://github.com/PyCQA/pycodestyle/pull/993
```
[the remark this answers is not in the document]
Comment (cdce8p) [reply astroid/__init__.py:184]: `pycodestyle` wasn't moved. So the changed link is actually correct: https://github.com/PyCQA/pycodestyle/pull/993
Didn't want to create a new PR just for that though.

[the remark this answers is not in the document]
Comment (Pierre-Sassoulas) [reply astroid/__init__.py:184]: Huh, my bad.

**Comment by cdce8p** [comment]
Then let's go ahead with this one then?

> I have no problem in doing a "early 2000 firefox" and release a lot of major each time we want to break things.

I think we should still try to bundle breaking changes at least as far as we can. To make it easier for us though, we could release alpha / beta versions for each new "area" of breaking changes. Those could then be pinned in `pylint 3.0` to make adopting them easier. That way we wouldn't have to resolve all issues at once.

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -26,8 +26,8 @@ repos:
     rev: v3.3.1
     hooks:
       - id: pyupgrade
-        exclude: tests/testdata
-        args: [--py37-plus]
+        exclude: astroid/__init__.py|tests/testdata
```
Comment (Pierre-Sassoulas) [inline .pre-commit-config.yaml:29]: Why must we do that ?

File: ChangeLog [hunk 3]
```diff
@@ -81,6 +81,9 @@ Release date: TBA
 
   Refs #1490
 
+* Remove support for Python 3.7.
+
```
Comment (Pierre-Sassoulas) [inline ChangeLog:85]:
```suggestion
* Remove support for Python 3.7.

Refs #2137 
```

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
So the plan is to:
- drop python 3.7 in astroid
- Move all the issues from astroid 2.16.0 in astroid 3.0, 
- Release astroid 3.0.0a0
- drop python 3.7 in pylint main branch
-  Release astroid 3.0.0 once it works in pylint and we're sure of the breaking changes we did
- Release pylint 3.0.0b1

Right ?

**Comment by DanielNoord** [comment]
Agree with both of you. The earlier we release an alpha we can spot mistakes I made with the current breaking changes. It is hard to know what is wrong or good with the minimal security of the astroid test suite currrently.

Release plan also sounds good!

**Comment by cdce8p** [comment]
> So the plan is to:
> 
> * drop python 3.7 in astroid
> * Move all the issues from astroid 2.16.0 in astroid 3.0,
> * Release astroid 3.0.0a0
> * drop python 3.7 in pylint main branch
> * Release astroid 3.0.0 once it works in pylint and we're sure of the breaking changes we did
> * Release pylint 3.0.0b1

I would modify that a bit:
- Drop Python 3.7 in astroid + pylint (can be done independently on one another - i.e. we don't need a new astroid release to merge https://github.com/pylint-dev/pylint/pull/8609)
- Move all the issues from astroid 2.16.0 in astroid 3.0
- Update changelog
- Release astroid 3.0.0a0
- Do additional breaking changes, cleanup, and fix bugs, e.g. remove `cachedproperty`, change TryExcept and TryFinally to just `Try` #1867, changes to the `Import` nodes, ...
- After each major change release new alpha for astroid
- For pylint 3.0 beta: Update pinned astroid version to use latest alpha release
- Once everything works and most/all of the changes are done -> release astroid 3.0
- ...
- While doing this: Continue to release new pylint 3.0 alpha or beta releases so we can gather feedback

**Review by DanielNoord — APPROVED** [review]
I would only argue that this should be at the top of our changelog for the `3.x` version. Rest LGTM! Thanks @cdce8p

**Review by DanielNoord — APPROVED** [review]

**Pull request merged**